### ISSUE TITLE
Non Deterministic Behaviour even after cudnn.deterministic = True and cudnn.benchmark=False 

### ISSUE THREAD

**Issue by Naman-ntc** [pytorch/pytorch#12207]
I was performing adversarial attacks with pytorch and I was not able to reproduce my experiments.
Finally I learned about non deterministic cudnn operations [Discussion Forum-1](https://discuss.pytorch.org/t/pytorch-doubletensor-operations-are-slow-while-floattensors-autograd-precision-is-low/26187) [Discussion Forum-2](https://discuss.pytorch.org/t/inconsistent-gradient-values-for-the-same-input/26179) and [final post](https://discuss.pytorch.org/t/nondeterminism-even-when-setting-all-seeds-0-workers-and-cudnn-deterministic/26080/4)
You can simply try to reproduce output from the following script (Ubuntu OS, Nvidia 1080 Ti GPU). 

If i change my model to a simple resnet (and appropirately the target and loss function) then it results are deterministic. (So I guess maybe Upsample is the culprit, I don't know!!). 
Here is my complete script : 
```                                                                                                                                                                                       
 import torch
 import torch.nn as nn
 torch.backends.cudnn.deterministic = True
 torch.backends.cudnn.benchmark = False
  
 class simple(nn.Module):
         def __init__(self):
                 super(simple, self).__init__()
                 self.seq = nn.Sequential(
                                 nn.Conv2d(3,64,7,2,3),
                                 nn.Conv2d(64,64,3,1,1),
                                 nn.MaxPool2d(2,2),
                                 nn.Conv2d(64,64,3,1,1),
                                 nn.MaxPool2d(2,2),
                                 nn.Upsample(scale_factor = 2),
                                 nn.Conv2d(64,64,3,1,1),
                                 nn.Upsample(scale_factor = 2),
                                 nn.Conv2d(64,1,1)
                         )

         def forward(self, x):
                 return self.seq(x)
 
 gpuid = 0
 
 ###
 model = simple().to(gpuid)
 lossfn = nn.MSELoss()
 img = torch.randn(1,3,256,256).float().to(gpuid)
 ###
 
 out = model(img)
 target = torch.randn(1,1,128,128).to(gpuid)
 print(out.norm())
 
 def test():
         adv = torch.zeros_like(img).to(gpuid)
 
         for i in range(100):
                 inpcopy = torch.clamp(img + adv, 0, 1)
                 inpcopy.requires_grad = True
                 loss = lossfn(model(inpcopy), target)
                 loss.backward()
                 adv = torch.clamp(torch.sign(inpcopy.grad) + adv, 0, 1)

         inpcopy = torch.clamp(img + adv, 0, 1)
         out = model(inpcopy)
         outNorm = float(out.norm())
         print(outNorm)
 

 while True:
     test() # Since there is no randomness the output perturbed image norm must be consistent
```

Can someone please look into it. And if it is problem of Upsample is there some hacky solution out of it??

**Comment by t-vi** [comment]
I don't think there is much to do, and FGSA is intentionally discontinuous (due to the sign). What you could do is determine the digit where it is unstable empirically and round that away before taking the sign.

That said, the reason is likely that the backward uses the cuda function `atomicAdd`, which is non-deterministic. For a factor of two, it would seem that
```python
class MyUpsample2(nn.Module):
    def forward(self, x):
        return x[:, :, :, None, :, None].expand(-1, -1, -1, 2, -1, 2).reshape(x.size(0), x.size(1), x.size(2)*2, x.size(3)*2)
```
makes this deterministic.

I'll probably add a note to https://github.com/pytorch/pytorch/blob/master/docs/source/notes/randomness.rst about atomic add (Users include: index_add, scatter_add, bincount, lossctc in forward, embedding/embedding_bag, all sorts of pooling, padding, sampling in backward, possibly also sparse to dense or coalescing, but I didn't check).

**Comment by Naman-ntc** [comment]
Is pooling non deterministic even after setting `torch.backends.cudnn.deterministic = True`. Also for a simple resnet18 the about script was pretty consistent..

**Comment by t-vi** [comment]
Pooling (and all of the above) isn't cudnn (or it is the non-cudnn version) , `atomicAdd` is a plain cuda function. We don't have a deterministic switch for the rest of PyTorch.

**Comment by Naman-ntc** [comment]
> Pooling (and all of the above) isn't cudnn (or it is the non-cudnn version) , `atomicAdd` is a plain cuda function. We don't have a deterministic switch for the rest of PyTorch.

Oh alright, I don't have much knowledge about cuda. Thanks!

**Comment by Naman-ntc** [comment]
@t-vi If you intend to update the docs, also maybe adding a note near the respective modules (similar to Conv module docs) makes sense. Atleast for common layers (such as upsample, or pooling (I am not 100% clear how that works but still yeah))

**Comment by Naman-ntc** [comment]
Hi @t-vi I guess the Bilinear Upsample layer will also be non deterministic, is there any fix for it similar to nearest upsampling??

**Comment by t-vi** [comment]
It's probably a bit more elaborate, but you could use F.unfold/F.fold to get the datapoints and apply the interpolation on that. Then you just have to merge the interpolated and original points.
Or you might just write the custom backward kernel for integral upsampling.

**Comment by Naman-ntc** [comment]
Seems [#12217](https://github.com/pytorch/pytorch/pull/12217) does try to clear this up. Closing this.

**Comment by t-vi** [comment]
I wonder whether there would be enough interest in at least "locally" deterministic behaviour (run on the same machine twice, get the same results) to offer a deterministic set of kernels. It's going to be quite a bit of work, but I would think that many ops could be helped.

**Comment by hazirbas** [comment]
I observed the same issue with MaxUnpool2d layer and I believe it is quite necessary to have determinism, at least for research purpose. Otherwise results on different experiments are not comparable.

**Comment by Naman-ntc** [comment]
@hazirbas hi yeah but mostly this won't be an issue for normal problems like recognition and types. It's due to non differentiable nature of gradient sign method that this becomes visible otherwise it's not big of a deap i guess

**Comment by simonhessner** [comment]
@t-vi 
> ```python
> class MyUpsample2(nn.Module):
>     def forward(self, x):
>         return x[:, :, :, None, :, None].expand(-1, -1, -1, 2, -1, 2).reshape(x.size(0), x.size(1), x.size(2)*2, x.size(3)*2)
> ```
> 
> makes this deterministic.
> 

Thank you very much! This solves the problem with the non-deterministic behaviour. It even runs faster in my tests.
Does this produce the same result as the Upsampling layer or are there cases where your method would fail?

**Comment by apple2373** [comment]
I found this post helpful. Thanks! 

I made the layer that is flexible in terms of upsampling size. 
```
import torch.nn as nn
class UpsampleDeterministic(nn.Module):
    def __init__(self,upscale=2):
        super(UpsampleDeterministic, self).__init__()
        self.upscale = upscale

    def forward(self, x):
        '''
        x: 4-dim tensor. shape is (batch,channel,h,w)
        output: 4-dim tensor. shape is (batch,channel,self.scale*h,self.scale*w)
        '''
        return x[:, :, :, None, :, None].expand(-1, -1, -1, self.upscale, -1, self.upscale).reshape(x.size(0), x.size(1), x.size(2)*self.upscale, x.size(3)*self.upscale)
```

**Comment by alimafukuro** [comment]
> I found this post helpful. Thanks!
> 
> I made a layer that is flexible in terms of upsampling size.
> 
> ```
> import torch.nn as nn
> class UpsampleDeterministic(nn.Module):
>     def __init__(self,upscale=2):
>         super(UpsampleDeterministic, self).__init__()
>         self.upscale = upscale
> 
>     def forward(self, x):
>         '''
>         x: 4-dim tensor. shape is (batch,channel,h,w)
>         output: 4-dim tensor. shape is (batch,channel,self.upscale*h,self.upscale*w)
>         '''
>         return x[:, :, :, None, :, None].expand(-1, -1, -1, self.upscale, -1, self.upscale).reshape(x.size(0), x.size(1), x.size(2)*self.upscale, x.size(3)*self.upscale)
> ```
> 
> or just a function
> 
> ```
> def upsample_deterministic(x,upscale):
>     return x[:, :, :, None, :, None].expand(-1, -1, -1, upscale, -1, upscale).reshape(x.size(0), x.size(1), x.size(2)*upscale, x.size(3)*upscale)
> ```

emmm, I thought if you set the 'mode = nearest' of 'torch.nn.UpSample', then it would become deterministic. I mean it's not the fault of original function, but the specific interpolation algorithm.
